MINIFICPP-1401 Read certificates from the Windows system store

fgerlits opened the pull request:
https://issues.apache.org/jira/browse/MINIFICPP-1401

If the (new) `nifi.security.use.system.cert.store` property is set to true, then read client and server certificates from the Windows system store if no certificate file is given in the properties.  On Linux/Mac, only server certificates are supported, and even that is untested.

By default, we use `LocalMachine/ROOT` ("Trusted Root Certification Authorities") as the server cert store and `LocalMachine/MY` ("Personal") as the client cert store, but these can be overridden by setting the
`nifi.security.windows.cert.store.location` (default: `LocalMachine`)
`nifi.security.windows.server.cert.store` (default: `ROOT`)
`nifi.security.windows.client.cert.store` (default: `MY`)
properties.

If `nifi.security.windows.client.cert.cn` is given a non-empty value, then the client certificate will only be accepted if it has this CN (in the Subject).

`nifi.security.windows.client.cert.key.usage` can contain a list of comma-separated Extended Key Usage strings; the default is "Client Authentication".  The client certificate will only be accepted if its Extended Key Usage contains all these key usages (it is allowed to contain more).  

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

fgerlits commented on libminifi/src/utils/tls/WindowsCertStoreLocation.cpp:
```diff
@@ -0,0 +1,85 @@
+
+namespace {
+
+constexpr std::array<std::pair<const char*, DWORD>, 8> SYSTEM_STORE_LOCATIONS{{
```
There is a reason, but not a very good one.  :)

Originally both were `std::pair`; I prefer that as it means less code, and it is more similar to `std::map`.  But the constructor of `std::pair` is not constexpr before C++14, and AppleClang is strict about that, so I had to change `std::pair` to a custom struct in code which is compiled on MacOS.  `WindowsCertStoreLocation` is only compiled on Windows, which is on C++14 already, so I could use `std::pair` here.

lordgamez replied:
I see, thanks for the explanation :)

adamdebreceni commented on libminifi/include/utils/tls/CertificateUtils.h:
```diff
@@ -0,0 +1,61 @@
+X509_unique_ptr convertWindowsCertificate(const PCCERT_CONTEXT certificate);
+
+// Returns nullptr if the certificate has no associated private key, or the private key could not be extracted
+EVP_PKEY_unique_ptr extractPrivateKey(const PCCERT_CONTEXT certificate);
```
top-level `const` has no effect here, a parameter passed by value being `const` is an implementation detail IMO and should be only in the definition
https://abseil.io/tips/109

fgerlits replied:
Thanks, fixed in https://github.com/apache/nifi-minifi-cpp/pull/947/commits/e63e9e6c64be860026b6756fd4ffce85ee3d7d8f.

adamdebreceni commented on libminifi/src/controllers/SSLContextService.cpp:
```diff
@@ -269,10 +505,8 @@ void SSLContextService::onEnable() {
   }
-  // load CA certificates
-  if (!getProperty(caCert.getName(), ca_certificate_)) {
-    logger_->log_error("Can not load CA certificate.");
```
is the log no longer necessary?

fgerlits replied:
It isn't an error any longer, because now there is a valid use case when the server certificate name is blank and `nifi.security.use.system.cert.store=true`.  In that case, we will try to find the server certificate in the system store.

szaszm commented on libminifi/src/utils/tls/ExtendedKeyUsage.cpp:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+bool operator<=(const ExtendedKeyUsage& left, const ExtendedKeyUsage& right) {
+  return (left.bits_ & right.bits_) == left.bits_;
+}
```
Take `A=1011` and `B=1100`. `A <= B` is false and `B <= A` is false. This should be impossible if this is meant to be used as a comparison operator. If this is meant to be some kind of arrow, then it needs a comment explaining the operation, or even better, a name.

adamdebreceni replied:
I think it is more of a "subset" relationship, agree that the operator is misleading

fgerlits replied:
Yes, it is a subset relation, which is a partial ordering.  `<=` seemed to me like a natural notation for this, but apparently it's not natural to anyone else. :)

I have changed it from `<=` to `isSubsetOf()`.

adamdebreceni commented on libminifi/include/utils/tls/WindowsCertStoreLocation.h:
```diff
@@ -0,0 +1,52 @@
+  DWORD getBitfieldValue() const { return location_bitfield_value_; }
+
+  static std::string defaultLocation();
+  static std::set<std::string> allowedLocations();
```
I think it is because `withAllowedValues` takes an `std::set`

fgerlits replied:
Yes, it is a subset relation, which is a partial ordering.  `<=` seemed like a natural notation for this, but apparently it's not natural to anyone else. :)

I have changed it from `<=` to `isSubsetOf()`.

adamdebreceni commented on libminifi/src/utils/tls/DistinguishedName.cpp:
```diff
@@ -0,0 +1,64 @@
+
+utils::optional<std::string> DistinguishedName::getCN() const {
+  const auto it = std::find_if(components_.begin(), components_.end(),
+      [](const std::string& component) { return component.substr(0, 3) == "CN="; });
```
I don't think there would be any allocation as 3 characters is within the (usual) 20(ish) SSO limit, but agree that `compare` is borderline more descriptive

fgerlits replied:
I have changed this to use `compare`.

szaszm replied:
Adam, you're right, I didn't think about SSO. From a readability standpoint, both are good IMO, so I'm fine with either. Thanks.

fgerlits commented on libminifi/src/utils/tls/DistinguishedName.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+DistinguishedName DistinguishedName::fromSlashSeparated(const std::string &slash_separated_components) {
+  return DistinguishedName{utils::StringUtils::split(slash_separated_components, "/")};
```
This is called zero to 2 or 3 times per heartbeat if `nifi.security.use.system.cert.store=true` is set and `nifi.security.windows.client.cert.cn` is non-empty.  The number of elements in `slash_separated_components` is also small, typically 2 to 4.

I think a better performance improvement would be to create the `SSLContextService` object used in the heartbeat only once at startup, rather than at each heartbeat.  I am planning to do that later in a separate PR.

szaszm replied:
That's a great idea, I'm fine with leaving it as is in the meantime. I think deployments with the most frequent heartbeat intervals are around 1/sec, so it should be fine in either case.

hunyadi-dev commented on libminifi/include/utils/tls/CertificateUtils.h:
```diff
@@ -0,0 +1,61 @@
+using EVP_PKEY_unique_ptr = std::unique_ptr<EVP_PKEY, EVP_PKEY_deleter>;
+
+struct X509_deleter {
+  void operator()(X509* cert) const { X509_free(cert); }
```
This is great to have as an utility class! Can we replace all usages of X509 free with this deleter? There are still usages in `SSLContextService.cpp` and `SourceInitiatedSubscriptionListener.cpp`.

fgerlits replied:
I have replaced `X509_free()` with `X509_unique_ptr` in `SourceInitiatedSubscriptionListener.cpp` (although this is part of the openwsman extension, which is disabled by default and does not compile at the moment -- I will fix that in a later PR.)

In `SSLContextService::addP12CertificateToSSLContext()`, I would have to replace
```c++
  const auto certs_guard = gsl::finally([pkey, cert, ca]() {
    EVP_PKEY_free(pkey);
    X509_free(cert);
    sk_X509_pop_free(ca, X509_free);
  });
```
with
```c++
  utils::tls::EVP_PKEY_unique_ptr pkey_ptr{pkey};
  utils::tls::X509_unique_ptr cert_ptr{cert};
  const auto ca_deleter = gsl::finally([ca] { sk_X509_pop_free(ca, X509_free); });
```
ie., introduce 3 deleter variables instead of 1.  Do you think that is worth it?

adamdebreceni replied:
we could wrap the `PKCS12_parse` to return an optional structure containing these smart pointers, unclear if worth the added complexity

fgerlits replied:
Yeah, if `PKCS12_parse()` was used anywhere else, I would do that, but for this one usage I don't think it's worth it.

hunyadi-dev replied:
>  Do you think that is worth it?

Yes, I think your change clears it up a lot.

fgerlits replied:
done: https://github.com/apache/nifi-minifi-cpp/pull/947/commits/6e90811fd2e8f4bf261d519e82dc6d87113fddeb

lordgamez commented:
Tested manually the possible configuration options on Windows 10. They all worked well,  I only have 2 comments:

1. While loading the server certificates, there is the possibility that a certificate might be loaded multiple times which results in an error `cert already in hash table`. As this is done for every heartbeat, this might flood the logs:
```
[2020-12-07 10:51:52.196] [org::apache::nifi::minifi::controllers::SSLContextService] [error] Failed to add server certificate /C=IE/O=Baltimore/OU=CyberTrust/CN=Baltimore CyberTrust Root to the SSL store; error: error:0BFFF065:x509 certificate routines:CRYPTO_internal:cert already in hash table
[2020-12-07 10:51:52.197] [org::apache::nifi::minifi::controllers::SSLContextService] [error] Failed to add server certificate /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert High Assurance EV Root CA to the SSL store; error: error:0BFFF065:x509 certificate routines:CRYPTO_internal:cert already in hash table
[2020-12-07 10:51:52.430] [org::apache::nifi::minifi::controllers::SSLContextService] [error] Failed to add server certificate /C=IE/O=Baltimore/OU=CyberTrust/CN=Baltimore CyberTrust Root to the SSL store; error: error:0BFFF065:x509 certificate routines:CRYPTO_internal:cert already in hash table
[2020-12-07 10:51:52.431] [org::apache::nifi::minifi::controllers::SSLContextService] [error] Failed to add server certificate /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert High Assurance EV Root CA to the SSL store; error: error:0BFFF065:x509 certificate routines:CRYPTO_internal:cert already in hash table
```
2. I encountered an error, that my private key was not exportable and the certificate could not be used because of this. As this is a requirement this should be noted in the documentation as well.
```
[2020-12-07 11:29:48.841] [org::apache::nifi::minifi::controllers::SSLContextService] [debug] Skipping client certificate /OU=NIFI/CN=minifi because it has no exportable private key
[2020-12-07 11:29:48.841] [org::apache::nifi::minifi::controllers::SSLContextService] [error] Could not find any suitable client certificate in sytem store LocalMachine/MY
```

fgerlits commented:
@lordgamez:

> Tested manually the possible configuration options on Windows 10. They all worked well, I only have 2 comments:
> 
>     1. While loading the server certificates, there is the possibility that a certificate might be loaded multiple times which results in an error `cert already in hash table`. As this is done for every heartbeat, this might flood the logs:

I have moved this log to debug level in https://github.com/apache/nifi-minifi-cpp/pull/947/commits/a4ee1fcdd27be870470af9d0a754214197dd8fcc; please retest.

>     2. I encountered an error, that my private key was not exportable and the certificate could not be used because of this. As this is a requirement this should be noted in the documentation as well.

We have discussed this with the user; they agreed to use exportable client certificates.  I have added a comment about this to the Jira, and we will include it in the documentation.

lordgamez commented on libminifi/src/controllers/SSLContextService.cpp:
```diff
@@ -128,16 +149,234 @@ bool SSLContextService::configure_ssl_context(SSL_CTX *ctx) {
+  int success = X509_STORE_add_cert(ssl_store, x509_cert.get());
+  if (success == 1) {
+    logger_->log_debug("Added server certificate %s from the system store to the SSL store", x509_cert->name);
+  } else if (ERR_peek_last_error() == X509_R_CERT_ALREADY_IN_HASH_TABLE) {
+    logger_->log_debug("Ignoring duplicate server certificate %s", x509_cert->name);
+  } else {
+    logger_->log_error("Failed to add server certificate %s to the SSL store; error: %s", x509_cert->name, getLatestOpenSSLErrorString());
+  }
```
I tested the latest change, but unfortunately it does not change the logging. The `X509_R_CERT_ALREADY_IN_HASH_TABLE` we are trying to check has the value `101` but after I log the `ERR_peek_last_error` it returns `201322597` so it is not the error code we were looking for. As checking further the error code in binary is `1011 1111 1111 1111 0000 0110 0101` where the last byte's value is actually `101`.

According to the [OpenSSL documentation](https://www.openssl.org/docs/man1.0.2/man3/ERR_GET_FUNC.html) the returned value of `ERR_peek_last_error` consists of a library number, function code and reason code and there are separate macros defined to retrieve needs. We need to use `ERR_GET_REASON` to retrieve the error code.

I tested the following change and it seems to be working.

```suggestion
  } else {
    auto err = ERR_peek_last_error();
    if (ERR_GET_REASON(err) == X509_R_CERT_ALREADY_IN_HASH_TABLE) {
      logger_->log_debug("Ignoring duplicate server certificate %s", x509_cert->name);
    } else {
      logger_->log_error("Failed to add server certificate %s to the SSL store; error: %s", x509_cert->name, getLatestOpenSSLErrorString());
    }
  }
```

fgerlits replied:
Thanks for the fix!  I have rewritten it slightly to avoid nested `if` statements.  Can you retest it, please?

lordgamez replied:
I retested it, it's working

fgerlits commented on libminifi/src/utils/tls/ExtendedKeyUsage.cpp:
```diff
@@ -0,0 +1,104 @@
+  const char* key;
+  uint8_t value;
+};
+constexpr std::array<KeyValuePair, 6> EXT_KEY_USAGE_NAME_TO_BIT_POS{{
+    KeyValuePair{"Server Authentication", 1},
+    KeyValuePair{"Client Authentication", 2},
+    KeyValuePair{"Code Signing", 3},
+    KeyValuePair{"Secure Email", 4},
+    KeyValuePair{"Time Stamping", 8},
+    KeyValuePair{"OCSP Signing", 9}
+}};
```
I have added a comment with a link to the RFC

fgerlits commented on libminifi/src/utils/tls/ExtendedKeyUsage.cpp:
```diff
@@ -0,0 +1,104 @@
+  for (int i = 0; i < num_oids; ++i) {
+    const ASN1_OBJECT* const oid = sk_ASN1_OBJECT_value(&key_usage_asn1, i);
+    assert(oid && oid->length > 0);
+    const unsigned char bit_pos = oid->data[oid->length - 1];
```
I have made some code improvements: rename variable, add comment, use `std::bitset`, which make this code less incomprehensible, I hope.

szaszm replied:
My problem is that I couldn't find a source that describes the structure of extended key usage asn1 objects in detail. I don't understand why we index to the last element (byte?) of the object and why it contains the bit position that we use below.

fgerlits replied:
In the RFC link https://tools.ietf.org/html/rfc5280#section-4.2.1.12 you can see that all the OIDs are the same except for the last byte.  The ASN.1 format is a list of OIDs; in this class we represent it as a bitfield, that is not a standard thing.

szaszm replied:
Ah, I see. I had some trouble reading ASN, but I think I see what you mean now: id-kp is the same and the last element (byte?) is different.

fgerlits commented on libminifi/src/controllers/SSLContextService.cpp:
```diff
@@ -128,16 +149,239 @@ bool SSLContextService::configure_ssl_context(SSL_CTX *ctx) {
+    }
+  }
+
+  utils::tls::EXTENDED_KEY_USAGE_unique_ptr key_usage{static_cast<EXTENDED_KEY_USAGE*>(X509_get_ext_d2i(x509_cert.get(), NID_ext_key_usage, nullptr, nullptr))};
+  if (!key_usage) {
+    logger_->log_error("Skipping client certificate %s because it has no extended key usage", x509_cert->name);
+    return false;
+  }
+
+  if (!(client_cert_key_usage_.isSubsetOf(utils::tls::ExtendedKeyUsage{*key_usage}))) {
+    logger_->log_debug("Skipping client certificate %s because its extended key usage set does not contain all usages specified in %s",
+                       x509_cert->name, Configuration::nifi_security_windows_client_cert_key_usage);
+    return false;
+  }
```
It would, if (1) I had known about it, and (2) this function was available in libressl.  This is a relatively new addition in openssl: https://github.com/openssl/openssl/commit/063f1f0c693a10aab6a7227df15d4120ed824856 and is not in libressl, yet: https://github.com/libressl-portable/openbsd/blob/master/src/lib/libcrypto/x509/x509v3.h#L789

I think we should ditch libressl and switch to openssl when 3.0.0 is released, as that will (finally) be under an Apache license.

szaszm replied:
Oh, that explains it, thanks. The switching sounds like a mid to longterm objective, so I'm marking this resolved now.
